usb/manual/ssdt.md

```diff
@@ -0,0 +1,165 @@
+## SSDT method
+
+A little bit of history worth to mention:
+
+* Thanks to [Gengik84](https://www.macos86.it/profile/1-gengik84/) and the [macos86.it community](https://www.macos86.it/topic/9-mappatura-porte-usb/) for developing and testing this quick and easy method, the original post dates back to 2018, even though some folks [stole it later in 2019](https://www.olarila.com/topic/6181-guide-native-usb-fix-for-desktops-no-injectorkext-required-skylake/) without mentioning any credits.
+
+## Prerequisites:
+
+* This guide assumes that the user knows how to use MaciASL, IORegistryExplorer and ProperTree. We'll use a sample motherboard, the ASUS Z370 Prime A II where the USB devices are described in the `SSDT_xh-rvp08.aml`. 
+* Use OpenCore's `SysReport` feature to look where your OEM put the definitions for your USB devices in the ACPI tables (if none found in the SSDTs, your last chance should be your DSDT)
+
+Before any modification, write down the `Length` value which is located in the beginning of the file, as shown: 
+ 
+![](../../images/post-install/usb-md/usb-map-ssdt-table-lenght.png)  
+
+The unmodified, original ACPI table has a lenght of `2835`, this is going to be useful later in the guide, when we're going to tell OpenCore to avoid loading that table in favour of the new one we're going to modify.
+
+## Adding the method to the original SSDT
+
+Let's start by adding the `GENG` method that can be found in MaciASL's Patch menu or this [link](https://github.com/1alessandro1/OpenCore-Post-Install/blob/master/extra-files/UsbConnectorTypePatch.txt) copy all those lines present in the .txt to the clipboard.
```

Comment:
This needs to be adjusted to cover if it's the DSDT, and adjust the controller path in the patch.

Reply:
For now dortania does not recommend to inject a custom DSDT everywhere in the guides, so I omitted that part

Reply:
@1alessandro1 We do not mean inject a new DSDT, but handle cases where the maps are in the DSDT. ie. disable the RHUB defined in the DSDT and rebuild it in a new custom DSDT

For reference, my X299 board has 1 XHCI controller defined in the DSDT and 2 other XHCI controllers handled in SSDTs. Simply disabling the RHUB for all 3 and rebuilding would allow you to have 1 big map. This would also allow you to not drop SSDTs as well

Reply:
@khronokernel feel free to contribute in this PR or make a new PR from my fork with the RHUB method, but I find it a little bit awkward. 

Or somebody can contribute to add a specific part for HEDT which requires special treatment later on

Reply:
@1alessandro1 This is not just an HEDT issue, this affect standard desktops as well as laptops. Yes you prefixed saying you don't support laptops however that's a big issue you're omiting by not mentioning it.

Regarding adapting, we mainly discussed supporting this PR specifically as we couldn't justify adjusting the guide for SSDT mapping ourselves . From our perspective, the PR as-is is not something we'd want to merge into dortania, we'd prefer a wider scope guide that can handle more cases as it is possible and fairly trivial to accomplish.

If you're not interested, you can simply mark this PR as a draft and leave it for a bit for someone else to pick up

Reply:
> Yes you prefixed saying you don't support laptops however that's a big issue you're omiting by not mentioning it.

I'm only following what you said in the guides. I cannot recommend to inject where everywhere is mentioned *not* to.

Reply:
Again, we are ***not*** recommending injecting a DSDT, we are referencing the RHUB disabling and creating a new map based of it

Reply:
Basically https://github.com/dortania/Getting-Started-With-ACPI/blob/master/extra-files/compiled/SSDT-RHUB.aml except we specify the ports too.

Reply:
Goodmoring @khronokernel  and @dhinakg, yesterday evening we and @dreamwhite found a clever way to map usb ports present in the DSDT without having to drop tables or having to load SSDT-RHUB mentioned above.

Basically by hot patching the `_UPC` method to `XUPC` we managed to define that method in an external SSDT-USB, like this one attached below. 

Let us know your opinions.
[Archive.zip](https://github.com/dortania/OpenCore-Post-Install/files/6136799/Archive.zip)

Reply:
Here's a demonstration from ioreg, where HS01 (originally defined as 0xff) is now correctly named as 0x03 since that port is a 3.0 port with retrocompatible USB 2.0 personality
![image](https://user-images.githubusercontent.com/46293832/111071600-35c6fb00-84d7-11eb-857b-99b375db1885.png)

Reply:
Here instead is an ACPI log extracted from dmesg with those hotpatches plus the custom SSDT table, as you can see, there are no errors related to USB and such

[AppleACPIPlatform-log.txt](https://github.com/dortania/OpenCore-Post-Install/files/6136808/AppleACPIPlatform-log.txt)

Reply:
Patches are generally iffy to begin with and we don't recommend using them if they're not needed. Writing generic patches is incredibly hard to do as due to how AML works, it's not always something you can take care of with a simple find and replace. For example, _OSI to XOSI can match against _SB_.OSIN, although we consider it a necessary evil as there hasn't any good option (currently, we're actively exploring options) that also isn't too complicated for most users.

To take this specific patch as an example, it will match to `_UPC` followed by a null byte, but it could also match to `_.UPC` followed by a null byte. Although this looks invalid (as names are padded with underscores so that they will always be four characters, and the above is three), you have to take into account AML opcodes that may raise issues. 

In addition to this, we also know that many firmwares have had a bad track record at being compliant to the ACPI spec. If a firmware defines a _UPC method for other purposes and gets XUPC instead, anywhere it is referenced in the ACPI will error + anything external (although unlikely) that is expecting _UPC will also error.

For these reasons, we'd like to avoid patches when possible. 

Even if patches were not a mess though, the implementation has some problems:
1. This patch assumes MethodFlags are always 0. Adjusting the patch to use a mask would result in it potentially matching other names.
2. If you don't include MethodFlags in find, then you would have to add bytes to the left to restrict matches, and PkgLength is of variable length.

Also this method of SSDT mapping has inherent problems in itself that I have realized today: any port that is special cased is going to be a mess. For example, if a port's definition is wrapped in if statements, that is going to suddenly make the SSDT more clunky. The "recreating RHUB" method does avoid this, however. Why not do this instead?

Due to this, we'd like to stick with kext mapping for now, as it has much less drawbacks and different casing that mapping via an SSDT. We expect to have major changes in USB mapping soon anyway that will require a rewrite, or at least major changes.

Reply:
> Patches are generally iffy to begin with and we don't recommend using them if they're not needed. Writing generic patches is incredibly hard to do as due to how AML works, it's not always something you can take care of with a simple find and replace. For example, _OSI to XOSI can match against _SB_.OSIN, although we consider it a necessary evil as there hasn't any good option (currently, we're actively exploring options) that also isn't too complicated for most users.
> 
> To take this specific patch as an example, it will match to `_UPC` followed by a null byte, but it could also match to `_.UPC` followed by a null byte. Although this looks invalid (as names are padded with underscores so that they will always be four characters, and the above is three), you have to take into account AML opcodes that may raise issues.
> 
> In addition to this, we also know that many firmwares have had a bad track record at being compliant to the ACPI spec. If a firmware defines a _UPC method for other purposes and gets XUPC instead, anywhere it is referenced in the ACPI will error + anything external (although unlikely) that is expecting _UPC will also error.

Dear @dhinakg good morning. first of all thank you so much for replying us. I perfectly got your point regarding the `_UPC` method invalidation but I have a few doubts that some firmwares may use `_UPC` for other purposes.
At least, according to [Section 9.14 of ACPI Specification manual](https://uefi.org/sites/default/files/resources/ACPI_6_3_May16.pdf#page=672) this method, despite is optional, is necessary for specifying custom USB Port Capabilites. Eventually may you gently provide some resources of monkey-written firmwares so we can try to adjust the patching mechanism?


>     1. This patch assumes MethodFlags are always 0. Adjusting the patch to use a mask would result in it potentially matching other names.

Regarding this point, I've thought about it these days and probably this entire PR should be moved onto `Getting Started with ACPI` or at least must have a section on it which explains why we should use some ACPI patches and how to obtain them and how to be sure that only the wanted stuff gets patched (you know, you add a Find pattern which must match only one thing. How to be sure about that? Open up DSDT.aml using Hex Fiend and look for the pattern).
Probably, you'd like that first [dortania's bugtracker issue #9](https://github.com/dortania/bugtracker/issues/9) gets completed, then we can start talking about ACPI renames. Don't you think so?
If yes, I want you to notice that I'm """writing""" a guide for patching the battery fields, both via static-patching (using DSDT.aml) and hot-patching (by using a blank SSDT-BATT.aml), and the hot-patching criteria is always the same. Let me know whether you're interested and eventually we may discuss on that later.

> Also this method of SSDT mapping has inherent problems in itself that I have realized today: any port that is special cased is going to be a mess. For example, if a port's definition is wrapped in if statements, that is going to suddenly make the SSDT more clunky.

I'm feeling a little bit confused about this statement. I mean, are you talking about the depicted below picture?
![image](https://user-images.githubusercontent.com/17933708/111129504-7033a500-8576-11eb-8ba0-7cf81414ee1f.png)

If yes, from what I've experienced by patching USB using the OEM SSDT Method (e.g. `xh_rvp08`), theorically there's no need to worry about that especially if we can first confirm that each port is working (or at least detected) without mapping. Then we can just specify the `_UPC` method for each port as I did with @1alessandro1's custom USB SSDT.   
If no, gently can you explain me better what are you talking about then? 😅

> Due to this, we'd like to stick with kext mapping for now, as it has much less drawbacks and different casing that mapping via an SSDT. We expect to have major changes in USB mapping soon anyway that will require a rewrite, or at least major changes.

That's a fair point but if so, why does @coprnewt in his USBMap README.md states that the kext approach is less clean than the ACPI method? Can you please give more detailed information about that?

> We can actually leverage this template to create our own injector kext that maps the ports on our motherboard!  This is the approach we'll use for this guide, and the approach that USBMap uses.

> **Pros:**

> * Uses the same approach Apple does
> * No extra code to execute
> * Can be mostly automated

**Cons:**

> * May not always be Apple's approach
> * Not as clean as an ACPI-only solution

Anyways, despite this topic is still hot and must have more attention on what we write and how we write, as this may lead users to confusion, I'd want to thank you for your patience and for your time.
So do I expect having major changes in USB mapping soon, so we can at least have more clarity on this topic.

Again, thank you so much and have a nice day